rfc/0078-zone-aware-replica-reads.md

```diff
@@ -0,0 +1,369 @@
+
+## Selecting Server Group for Connection
+
+To allow user to select preferred server groups, new setter/property for
+`ClusterOptions` must be introduced:
+
+```
+class ClusterOptions {
+    // ...
+    ClusterOptions preferredServerGroup(String serverGroupName);
+    // ...
+}
+```
+
+Where `serverGroupName` matches name of the group as it seen in Admin UI of the
+server and in cluster configuration JSON.
+
+In the case when, at the time of the request, there is no group with the given
+name, the SDK should still try to filter the nodes and return a `102
+DocumentUnretrievable` error because the result set is empty.
+
+## Selecting Read Preference for Operations
+
+New enumeration should be defined for expressing different strategies for the
+read replica APIs:
+
+```
+enum ReadPreference {
+    NO_PREFERENCE,
+    SELECTED_SERVER_GROUP,
+};
+```
+
+Each of the operation should get new option, with default value of
+`ReadPreference::NO_PREFERENCE`.
+
+```
+class GetAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class GetAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAllReplicasOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+```
+class LookupInAnyReplicaOptions {
+    // ...
+    ReadPreference readPreference { ReadPreference::NO_PREFERENCE };
+    // ...
+};
+```
+
+In all failure cases, when the SDK cannot handle operation, it must return `102
+DocumentUnretrievable` with human-readable explanation whenever it is possible
+that explains the details. This includes cases when the preferred server group
+was not selected using the `ClusterOptions::preferredServerGroup()` call.
+
+## Transactions
+
+Inside transaction closure, the replica read API takes simplified form, and
+always read from local group, which implies that the SDK will always throw `102
+DocumentUnretrievable` error for servers that do not expose `serverGroup`
+property in cluster configuration.
+
+```
+class TransactionAttemptContext {
+  // ...
+  public TransactionGetResult
+  getReplicaFromPreferredServerGroup(Collection collection, String id,
+                                     TransactionGetReplicaOptions options);
+  // ...
+}
+```
+
+`TransactionGetReplicaOptions` here does not have any specific options except
+common with `get()` method. The user cannot select read preference.
+
+```
+class TransactionGetReplicaOptions {
+    Transcoder transcoder;
+}
+```
+
+The method might throw the following errors:
+
+* `101 DocumentNotFound`, when the server returns KV Code `0x01 ENOENT` for
+  *all* requests.
+
+* `102 DocumentUnretrievable`, when the SDK finds that there are nodes in local
+  group, or there is no group available with the name selected in connection
+  options.
```

Comment:
This might be inconsistent with the non-transactional `GetAnyReplica` behaviour. The [error handling RFC](https://github.com/couchbaselabs/sdk-rfcs/blob/master/rfc/0058-error-handling.md#102-documentunretrievable) suggests that `DocumentUnretrievable` is the error to throw (not `DocumentNotFound` as suggested here) when all `GetReplica` requests fail with `0x01 ENOENT`:
> ### 102 DocumentUnretrievable
> Raised When
> *  In getAnyReplica, the getAllReplicas returns an empty stream because all the individual errors are dropped (i.e. all returned a DocumentNotFound)